src/main/java/com/startstepszalando/ecommerceshop/user/controller/UserController.java

```diff
@@ -84,14 +94,53 @@ public ResponseEntity<?> registerUser(
     public ResponseEntity<?> login(@Valid @RequestBody UserLoginRequest loginRequest) {
-        Authentication authentication = authenticationManager.authenticate(
-                new UsernamePasswordAuthenticationToken(loginRequest.getEmail(), loginRequest.getPassword())
-        );
+        Authentication authentication = authenticationManager
+                .authenticate(new UsernamePasswordAuthenticationToken(loginRequest.getEmail(), loginRequest.getPassword()));
 
         SecurityContextHolder.getContext().setAuthentication(authentication);
-        UserDetails userDetails = (UserDetails) authentication.getPrincipal();
-        final String jwt = jwtService.generateToken(userDetails);
 
-        return ResponseEntity.ok(new AuthenticationResponse(jwt, "User logged in successfully"));
+        UserImpl userDetails = (UserImpl) authentication.getPrincipal();
+
+        ResponseCookie jwtCookie = jwtService.generateJwtCookie(userDetails);
+
+        List<String> roles = userDetails.getAuthorities().stream()
+                .map(GrantedAuthority::getAuthority)
+                .toList();
+
+        RefreshToken refreshToken = refreshTokenService.createOrUpdateRefreshToken(userDetails.getId());
+
+        AccessToken responseBody = new AccessToken(userDetails.getUsername(), roles, jwtCookie.getValue(), refreshToken.getToken(), jwtCookie.getMaxAge().toString());
+
+        return ResponseEntity.ok().header(HttpHeaders.SET_COOKIE, jwtCookie.toString())
+                .body(responseBody);
+    }
+
+    @Operation(summary = "Refresh authentication token: Provide a valid refresh token to receive a new authentication token", responses = {
+            @ApiResponse(responseCode = "200", description = "Successfully refreshed authentication token",
+                    content = @Content(mediaType = "application/json",
+                            schema = @Schema(implementation = TokenRefreshResponse.class))),
+            @ApiResponse(responseCode = "404", description = "Refresh token is not in database",
+                    content = @Content(mediaType = "application/json",
+                            schema = @Schema(implementation = TokenValidationException.class))),
+            @ApiResponse(responseCode = "403", description = "Refresh token was expired",
+                    content = @Content(mediaType = "application/json",
+                            schema = @Schema(implementation = TokenRefreshException.class))),
+            @ApiResponse(responseCode = "500", description = "Internal server error",
+                    content = @Content(mediaType = "application/json",
+                            schema = @Schema(implementation = Exception.class)))
+    })
+    @PostMapping("/refreshtoken")
+    public ResponseEntity<?> refreshToken(@Valid @RequestBody TokenRefreshRequest request) {
+        String requestRefreshToken = request.getRefreshToken();
+
+        return refreshTokenService.findByToken(requestRefreshToken)
```

Comment:
move this logic to the service. We should have little to none business logic in our controller